File: Tutorial4-3ComponentBuilders/src/main/java/com/example/tutorial4_3componentbuilders/di/AppComponentWithBuilder.java
package com.example.tutorial4_3componentbuilders.di;

import android.app.Application;
import android.content.SharedPreferences;

import com.example.tutorial4_3componentbuilders.MainActivity;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;

@Singleton
@Component(modules = {AppModuleWithBuilder.class})

public interface AppComponentWithBuilder {
    void inject(MainActivity mainActivity);

    @Component.Builder
    interface Builder {

        AppComponentWithBuilder build();


        // @BindsInstance replaces Builder appModule(AppModule appModule)
        // And removes Constructor with Application AppModule(Application)

        @BindsInstance
        Builder application(Application application);
    }
}

File: Tutorial6-2ComponentFactoryVsBuilder/src/main/java/com/smarttoolfactory/tutorial6_2componentfactoryvsbuilder/di/AppComponentWithBuilder.kt
package com.smarttoolfactory.tutorial6_2componentfactoryvsbuilder.di

import android.app.Application
import android.content.SharedPreferences
import com.smarttoolfactory.tutorial6_2componentfactoryvsbuilder.MainActivity
import dagger.BindsInstance
import dagger.Component

@Component(modules = [AppModule::class])
interface AppComponentWithBuilder {

    fun inject(mainActivity: MainActivity)


    @Component.Builder
    interface Builder {

        fun build(): AppComponentWithBuilder

        // @BindsInstance replaces Builder appModule(AppModule appModule)
        // And removes Constructor with Application AppModule(Application)
        @BindsInstance
        fun application(application: Application): Builder

        @BindsInstance
        fun sharedPreferences(sharedPreferences: SharedPreferences): Builder

    }

}